lib/user_app_jwt/user_app_jwt.c

```diff
@@ -0,0 +1,758 @@
+#include <psa/crypto_extra.h>
+#include <psa/nrf_platform_key_ids.h>
+
+#include <cJSON.h>
+
+#include <user_app_jwt.h>
+#if defined(CONFIG_DATE_TIME)
+#include <date_time.h>
+#endif
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(user_app_jwt, CONFIG_USER_APP_JWT_LOG_LEVEL);
+
+/* Size of a UUID in words */
+#define UUID_BINARY_WORD_SZ (4)
+
+/* Size of a UUID in bytes */
+#define UUID_BINARY_BYTES_SZ (16)
+
+/* String size of a binary word encoded in hexadecimal */
+#define BINARY_WORD_STR_SZ (9)
+
+/* Size of an ECDSA signature */
+#define ECDSA_SHA_256_SIG_SZ (64)
+
+/* Size of an SHA 256 hash */
+#define ECDSA_SHA_256_HASH_SZ (32)
+
+/* Size of a public ECDSA key in raw binary format  */
+#define ECDSA_PUBLIC_KEY_SZ (65)
+
+/* Size of a public ECDSA key in DER format  */
+#define ECDSA_PUBLIC_KEY_DER_SZ (91)
+
+/* Macro to determine the size of a data encoded in base64 */
+#define BASE64_ENCODE_SZ(n) (((4 * n / 3) + 3) & ~3)
+
+/* Size of ECDSA signature encoded in base64 */
+#define B64_SIG_SZ (BASE64_ENCODE_SZ(ECDSA_SHA_256_SIG_SZ) + 1)
+
+/* Lenght of random bites added to JTI field */
+#define JWT_NONCE_BITS (128)
+
+/* Lenght of JTI field, this should be at least UUIDv4 size + size of board name */
+#define JWT_CLAIM_FILED_STR_LENGTH (64)
+
+/* lenght of uint64_t_max written as a string + 1 string termination character */
+#define JWT_EXPIRATION_STR_LENGTH (21)
+
+//---------------------------------------------------------------------------//
+static uint32_t swap_endian(uint32_t value)
+{
+	return ((value >> 24) & 0xFF) | ((value << 8) & 0xFF0000) | ((value >> 8) & 0xFF00) |
+	       ((value << 24) & 0xFF000000);
+}
+
+//---------------------------------------------------------------------------//
+static void base64_url_format(char *const base64_string)
+{
+	if (base64_string == NULL) {
+		return;
+	}
+
+	char *found = NULL;
+
+	/* replace '+' with "-" */
+	for (found = base64_string; (found = strchr(found, '+'));) {
+		*found = '-';
+	}
+
+	/* replace '/' with "_" */
+	for (found = base64_string; (found = strchr(found, '/'));) {
+		*found = '_';
+	}
+
+	/* remove padding '=' */
+	found = strchr(base64_string, '=');
+	if (found) {
+		*found = '\0';
+	}
+}
+
+//---------------------------------------------------------------------------//
+static int bytes_to_uuid_str(const uint32_t *uuid_words, const int32_t uuid_byte_len,
+			     char *uuid_str_out, const int32_t uuid_str_out_size)
+{
+
+	if ((NULL == uuid_words) || (uuid_byte_len < UUID_BINARY_WORD_SZ * 4) ||
+	    (NULL == uuid_str_out) || (uuid_str_out_size < NRF_UUID_V4_STR_LEN)) {
+		/* bad parameter */
+		return -EINVAL;
+	}
+	// this will hold 4 integer words in string format (string lenght 8 + 1)
+	char temp_str[UUID_BINARY_WORD_SZ][BINARY_WORD_STR_SZ] = {0};
+
+	// Transform integers to strings
+	for (int i = 0; i < UUID_BINARY_WORD_SZ; i++) {
+		// UUID byte endiannes is little endian first.
```

Comment:
Is this converting from LE to BE? From the comment it's not clear which one is little endian. Because the platform is little endian, I assume that the byte order in the string should be big endian. Anyway, I think `swap_endian()` should be replaced with Zephyr macro `sys_cpu_to_be32()`.